# Method
We used a merged VCF of 245,394 samples and 99,250,816 variants meeting the criteria of a population-specific allele frequency (AF) greater than 1% or a population-specific allele count (AC) greater than 100 in any computed ancestry subpopulations in chromosome 1–22 and X. We excluded 15,375 related samples flagged by All of Us using pre-computed kinship scores by Hail (PMID: 20926424) (relatedness_flagged_samples.tsv) of all pairs to identify the relatedness of all samples greater than 0.1, keeping only unrelated participants. Multi-allelic variants were split into biallelic variants and kept 230,019 unrelated samples. We performed reference-free phasing using BEAGLE v5.4 (PMID: 34478634) with window size 10–20 cM varied by chromosome length. We parallelized each run per chromosome across five separate AoU Cloud workspaces. Since each task in the analysis workflow required different compute resources, we created different cloud computing environments (Table 1). We used phased VCF files as input to local ancestry calling with FLARE (PMID: 36610402). We combined both HGDP and 1KGP to create a reference panel of 2,265 samples, which includes: 529 AFR (1KGP), 54 America (HGDP), 512 EAS (1KGP), 525 EUR (1KGP), 133 MIDDLE_EAST (HGDP), 512 SAS (1KGP) to ensure that each population consists of homogeneous samples similar to related studies (PMID: 36747613, 38374255). To improve accuracy suggested by the original developers of FLARE, we created population-specific VCF files using global ancestry prediction values precomputed with a random forest classifier based on principal components by AoU (PMID: 38374255). To manage computational loads, any subset containing more than 10,000 samples was further partitioned into multiple chunks of approximately 10,000 individuals each. For instance, on a typical chromosome, EAS (5,525 samples) required only one chunk, whereas EUR (128,302 samples) was split into 14 chunks. We ran FLARE separately on each chunked VCF. Once FLARE finished running on all chunks for a given chromosome, we merged those chunk-level output files back together to produce unified files that include all 230,019 unrelated AoU samples.

# Workflow
![FLARE Workflow](https://github.com/CAST-genomics/cast-workflows/raw/main/AoU-FLARE/workflow.png)

# Table 1: Task specific cloud computing resource
![Time and Cost Table](https://github.com/CAST-genomics/cast-workflows/raw/main/AoU-FLARE/Table1-time-cost.png)
